File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/analytics.rb
# frozen_string_literal: true

require_relative '../errors'
require_relative '../database'
require_relative 'campaigns'

module Handlers
  # Deterministic campaign analytics derived from state accumulated across
  # all prior stages: quests, npcs/factions, sessions, and inventory.
  module Analytics
    module_function

    def count(sql)
      row = Database.query(sql).first
      row ? row['count'].to_i : 0
    end

    def open_quests(campaign_id)
      count(<<~SQL)
        SELECT COUNT(*) AS count FROM campaign_quests
        WHERE campaign_id = #{Database.escape(campaign_id)} AND status = 'active';
      SQL
    end

    def friendly_npcs(campaign_id)
      count(<<~SQL)
        SELECT COUNT(*) AS count
        FROM campaign_npcs n
        JOIN campaign_factions f ON f.id = n.faction_id AND f.campaign_id = n.campaign_id
        WHERE n.campaign_id = #{Database.escape(campaign_id)} AND f.stance = 'friendly';
      SQL
    end

    def scheduled_sessions(campaign_id)
      count(<<~SQL)
        SELECT COUNT(*) AS count FROM campaign_sessions
        WHERE campaign_id = #{Database.escape(campaign_id)};
      SQL
    end

    def inventory_items(campaign_id)
      count(<<~SQL)
        SELECT COUNT(*) AS count FROM campaign_inventory
        WHERE campaign_id = #{Database.escape(campaign_id)};
      SQL
    end

    def has_characters(campaign_id)
      count(<<~SQL).positive?
        SELECT COUNT(*) AS count FROM campaign_characters
        WHERE campaign_id = #{Database.escape(campaign_id)};
      SQL
    end

    def readiness_score(open_quests, friendly_npcs, scheduled_sessions, inventory_items)
      raw = 10 + (open_quests * 20) + (friendly_npcs * 15) + (scheduled_sessions * 30) + (inventory_items * 10)
      raw.clamp(0, 100)
    end

    def summary(campaign_id, _body)
      Handlers::Campaigns.find_campaign(campaign_id)

      quests = open_quests(campaign_id)
      npcs = friendly_npcs(campaign_id)
      sessions = scheduled_sessions(campaign_id)
      items = inventory_items(campaign_id)

      [200, {
        campaign_id: campaign_id,
        readiness_score: readiness_score(quests, npcs, sessions, items),
        open_quests: quests,
        friendly_npcs: npcs,
        scheduled_sessions: sessions,
        inventory_items: items
      }]
    end

    def risk_report(campaign_id, body)
      campaign = Handlers::Campaigns.find_campaign(campaign_id)

      include_zeroes = body.key?('include_zeroes') ? body['include_zeroes'] : true
      raise HttpError.new(400, 'include_zeroes must be a boolean') unless [true, false].include?(include_zeroes)

      has_dm = campaign[:dm].is_a?(String) && !campaign[:dm].empty?
      has_chars = has_characters(campaign_id)
      has_next_session = scheduled_sessions(campaign_id).positive?
      has_active_quest = open_quests(campaign_id).positive?

      signals = {
        has_dm: has_dm,
        has_characters: has_chars,
        has_next_session: has_next_session,
        has_active_quest: has_active_quest
      }

      missing = []
      missing << 'dm' unless has_dm
      missing << 'characters' unless has_chars
      missing << 'next_session' unless has_next_session
      missing << 'active_quest' unless has_active_quest

      risk_level = case missing.size
                   when 0 then 'low'
                   when 1, 2 then 'medium'
                   else 'high'
                   end

      [200, {
        campaign_id: campaign_id,
        risk_level: risk_level,
        missing: include_zeroes ? missing : [],
        signals: signals
      }]
    end
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_rust-stdlib/src/analytics.rs
use crate::json::{
    escape_json, extract_bool, extract_int, extract_objects, extract_string, extract_top_array_content,
};
use crate::store::{campaign_exists, sql_escape, sqlite_query};

pub enum AnalyticsError {
    BadRequest,
    NotFound,
}

/// Return a deterministic campaign analytics summary.
///
/// GET /v1/campaigns/{id}/analytics/summary
pub fn analytics_summary(campaign_id: &str) -> Result<String, AnalyticsError> {
    if !campaign_exists(campaign_id).map_err(|_| AnalyticsError::BadRequest)? {
        return Err(AnalyticsError::NotFound);
    }

    let signals = campaign_signals(campaign_id)?;
    let true_count = signals.iter().filter(|&&v| v).count() as i64;
    let readiness_score = (25 + true_count * 15).min(100);

    let open_quests = count_open_quests(campaign_id)?;
    let friendly_npcs = count_friendly_npcs(campaign_id)?;
    let scheduled_sessions = count_scheduled_sessions(campaign_id)?;
    let inventory_items = count_inventory_items(campaign_id)?;

    Ok(format!(
        r#"{{"campaign_id":"{}","readiness_score":{},"open_quests":{},"friendly_npcs":{},"scheduled_sessions":{},"inventory_items":{}}}"#,
        escape_json(campaign_id),
        readiness_score,
        open_quests,
        friendly_npcs,
        scheduled_sessions,
        inventory_items
    ))
}

/// Return a deterministic maintenance risk report.
///
/// POST /v1/campaigns/{id}/analytics/risk-report
pub fn risk_report(campaign_id: &str, body: &str) -> Result<String, AnalyticsError> {
    if !campaign_exists(campaign_id).map_err(|_| AnalyticsError::BadRequest)? {
        return Err(AnalyticsError::NotFound);
    }

    let include_zeroes = extract_bool(body, "include_zeroes").unwrap_or(false);
    let signals = campaign_signals(campaign_id)?;
    let [has_dm, has_characters, has_next_session, has_active_quest] = signals;

    let mut missing = Vec::new();
    if !has_dm {
        missing.push("dm");
    }
    if !has_characters {
        missing.push("characters");
    }
    if !has_next_session {
        missing.push("next_session");
    }
    if !has_active_quest {
        missing.push("active_quest");
    }
    let core_missing = missing.len();

    if include_zeroes {
        if count_open_quests(campaign_id)? == 0 {
            missing.push("open_quests");
        }
        if count_friendly_npcs(campaign_id)? == 0 {
            missing.push("friendly_npcs");
        }
        if count_scheduled_sessions(campaign_id)? == 0 {
            missing.push("scheduled_sessions");
        }
        if count_inventory_items(campaign_id)? == 0 {
            missing.push("inventory_items");
        }
    }

    let risk_level = match core_missing {
        0 => "low",
        1 | 2 => "medium",
        _ => "high",
    };

    let missing_json = missing
        .iter()
        .map(|s| format!(r#""{}""#, s))
        .collect::<Vec<_>>()
        .join(",");

    Ok(format!(
        r#"{{"campaign_id":"{}","risk_level":"{}","missing":[{}],"signals":{{"has_dm":{},"has_characters":{},"has_next_session":{},"has_active_quest":{}}}}}"#,
        escape_json(campaign_id),
        risk_level,
        missing_json,
        bool_json(has_dm),
        bool_json(has_characters),
        bool_json(has_next_session),
        bool_json(has_active_quest)
    ))
}

fn campaign_signals(campaign_id: &str) -> Result<[bool; 4], AnalyticsError> {
    let dm = fetch_dm(campaign_id)?;
    Ok([
        !dm.is_empty(),
        count_characters(campaign_id)? > 0,
        count_scheduled_sessions(campaign_id)? > 0,
        count_active_quests(campaign_id)? > 0,
    ])
}

fn fetch_dm(campaign_id: &str) -> Result<String, AnalyticsError> {
    let sql = format!(
        "SELECT dm FROM campaigns WHERE id = '{}';",
        sql_escape(campaign_id)
    );
    let output = sqlite_query(&sql).map_err(|_| AnalyticsError::BadRequest)?;
    let content = extract_top_array_content(&output).ok_or(AnalyticsError::BadRequest)?;
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Err(AnalyticsError::NotFound);
    }
    extract_string(objects[0], "dm")
        .map(|s| s.to_string())
        .ok_or(AnalyticsError::BadRequest)
}

fn count_characters(campaign_id: &str) -> Result<i64, AnalyticsError> {
    count_rows(campaign_id, "characters", "1=1")
}

fn count_active_quests(campaign_id: &str) -> Result<i64, AnalyticsError> {
    count_rows(campaign_id, "quests", "status = 'active'")
}

fn count_open_quests(campaign_id: &str) -> Result<i64, AnalyticsError> {
    count_rows(campaign_id, "quests", "status != 'completed'")
}

fn count_friendly_npcs(campaign_id: &str) -> Result<i64, AnalyticsError> {
    count_rows(campaign_id, "npcs", "disposition > 0")
}

fn count_scheduled_sessions(campaign_id: &str) -> Result<i64, AnalyticsError> {
    count_rows(campaign_id, "campaign_sessions", "1=1")
}

fn count_inventory_items(campaign_id: &str) -> Result<i64, AnalyticsError> {
    let sql = format!(
        "SELECT COUNT(*) AS cnt FROM (SELECT item_slug FROM campaign_inventory WHERE campaign_id = '{}' AND quantity > 0 UNION SELECT item_slug FROM character_equipment WHERE campaign_id = '{}' AND quantity > 0);",
        sql_escape(campaign_id),
        sql_escape(campaign_id)
    );
    let output = sqlite_query(&sql).map_err(|_| AnalyticsError::BadRequest)?;
    let content = extract_top_array_content(&output).unwrap_or("");
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Ok(0);
    }
    extract_int(objects[0], "cnt").ok_or(AnalyticsError::BadRequest)
}

fn count_rows(campaign_id: &str, table: &str, predicate: &str) -> Result<i64, AnalyticsError> {
    let sql = format!(
        "SELECT COUNT(*) AS cnt FROM {} WHERE campaign_id = '{}' AND {};",
        table,
        sql_escape(campaign_id),
        predicate
    );
    let output = sqlite_query(&sql).map_err(|_| AnalyticsError::BadRequest)?;
    let content = extract_top_array_content(&output).unwrap_or("");
    let objects = extract_objects(content);
    if objects.is_empty() {
        return Ok(0);
    }
    extract_int(objects[0], "cnt").ok_or(AnalyticsError::BadRequest)
}

fn bool_json(v: bool) -> &'static str {
    if v {
        "true"
    } else {
        "false"
    }
}
